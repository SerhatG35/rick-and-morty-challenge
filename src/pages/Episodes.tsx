/* eslint-disable react-hooks/exhaustive-deps */
import { EpisodeTypes } from 'global';
import { useEffect, useState } from 'react';
import { useInView } from 'react-intersection-observer';
import EpisodeBox from 'src/components/EpisodeBox';
import { EpisodesGET } from 'src/service/rickAndMorty';
import { Center } from '../styles/styles';

const Episodes = () => {
  const [episodes, setEpisodes] = useState<EpisodeTypes[]>([]);
  const [page, setPage] = useState<number | null>(1);
  const [ref, inView] = useInView();

  const fetchEpisodes = async () => {
    if (page !== null) {
      const data = await EpisodesGET(page);
      setEpisodes((prev) => [...prev, ...data.results]);
      if (data.info.next !== null)
        setPage(Number(data.info.next.split('page=')[1]));
      else setPage(data.info.next);
    } else return;
  };
  useEffect(() => {
    fetchEpisodes();
  }, []);

  useEffect(() => {
    if (inView === true) fetchEpisodes();
  }, [inView]);

  return (
    <Center as="main" justifyContent="flex-start" width="70%">
      {episodes?.map((episode, index) => {
        if (index === episodes.length - 1)
          return (
            <EpisodeBox
              viewRef={ref}
              key={episode.id}
              episode={episode}
              data-testid="serhat"
            />
          );
        else
          return (
            <EpisodeBox
              key={episode.id}
              episode={episode}
              data-testid="serhat"
            />
          );
      })}
    </Center>
  );
};

export default Episodes;
