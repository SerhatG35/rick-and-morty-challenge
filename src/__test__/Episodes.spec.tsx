import { act, render, waitFor } from '@testing-library/react';
import Episodes from 'src/pages/Episodes';
import { useInView } from 'react-intersection-observer';

const useInViewMock = useInView as jest.Mock;

describe('Episode page tests', () => {
  test('generates a snapshot', async () => {
    useInViewMock.mockImplementation(() => [null, true]);

    const wrapper = render(<Episodes />);
    await waitFor(() => wrapper);

    expect(wrapper).toMatchSnapshot();
    wrapper.unmount();
  });
});
