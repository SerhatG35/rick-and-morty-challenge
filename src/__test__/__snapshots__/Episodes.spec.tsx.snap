// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Episode page tests generates a snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <main
        class="sc-bdnxRM dktVZV"
        width="70%"
      >
        <section
          class="sc-bdnxRM byZzmS"
          data-testid="episode-wrapper"
        >
          <h1
            class="sc-jSFjdj cYWCdr"
          >
            #90-Pilot
          </h1>
          <p
            class="sc-iCoGMd fFHScY"
          >
            This is the 01st episode in 01st session. It was aired on December 2, 2013. There are total of 1 featured characters in this episode
          </p>
          <div
            class="sc-bdnxRM dNzzIz"
          >
            <article
              class="sc-bdnxRM jDhRBX"
              scale="1.02"
              width="30%"
            >
              <span
                class=" lazy-load-image-background blur"
                style="display: inline-block; height: 50%; width: 100%;"
              >
                <img
                  alt="episode character"
                  height="50%"
                  src="https://rickandmortyapi.com/api/character/avatar/1.jpeg"
                  width="100%"
                />
              </span>
              <h2
                class="sc-eCApnc hQwxcN"
                color="#1f1f1f"
                font-style="normal"
              >
                Rick Sanchez
              </h2>
              <div
                class="sc-bdnxRM DhlHX"
              >
                <p
                  class="sc-iCoGMd gGpfF"
                >
                  From Earth (C-137), identifies as Male of Human race.
                </p>
                <p
                  class="sc-iCoGMd KOcH"
                >
                  Current Status: Alive
                </p>
              </div>
            </article>
          </div>
        </section>
        <section
          class="sc-bdnxRM byZzmS"
          data-testid="episode-wrapper"
        >
          <h1
            class="sc-jSFjdj cYWCdr"
          >
            #90-Pilot
          </h1>
          <p
            class="sc-iCoGMd fFHScY"
          >
            This is the 01st episode in 01st session. It was aired on December 2, 2013. There are total of 1 featured characters in this episode
          </p>
          <div
            class="sc-bdnxRM dNzzIz"
          >
            <article
              class="sc-bdnxRM jDhRBX"
              scale="1.02"
              width="30%"
            >
              <span
                class=" lazy-load-image-background blur"
                style="display: inline-block; height: 50%; width: 100%;"
              >
                <img
                  alt="episode character"
                  height="50%"
                  src="https://rickandmortyapi.com/api/character/avatar/1.jpeg"
                  width="100%"
                />
              </span>
              <h2
                class="sc-eCApnc hQwxcN"
                color="#1f1f1f"
                font-style="normal"
              >
                Rick Sanchez
              </h2>
              <div
                class="sc-bdnxRM DhlHX"
              >
                <p
                  class="sc-iCoGMd gGpfF"
                >
                  From Earth (C-137), identifies as Male of Human race.
                </p>
                <p
                  class="sc-iCoGMd KOcH"
                >
                  Current Status: Alive
                </p>
              </div>
            </article>
          </div>
        </section>
      </main>
    </div>
  </body>,
  "container": <div>
    <main
      class="sc-bdnxRM dktVZV"
      width="70%"
    >
      <section
        class="sc-bdnxRM byZzmS"
        data-testid="episode-wrapper"
      >
        <h1
          class="sc-jSFjdj cYWCdr"
        >
          #90-Pilot
        </h1>
        <p
          class="sc-iCoGMd fFHScY"
        >
          This is the 01st episode in 01st session. It was aired on December 2, 2013. There are total of 1 featured characters in this episode
        </p>
        <div
          class="sc-bdnxRM dNzzIz"
        >
          <article
            class="sc-bdnxRM jDhRBX"
            scale="1.02"
            width="30%"
          >
            <span
              class=" lazy-load-image-background blur"
              style="display: inline-block; height: 50%; width: 100%;"
            >
              <img
                alt="episode character"
                height="50%"
                src="https://rickandmortyapi.com/api/character/avatar/1.jpeg"
                width="100%"
              />
            </span>
            <h2
              class="sc-eCApnc hQwxcN"
              color="#1f1f1f"
              font-style="normal"
            >
              Rick Sanchez
            </h2>
            <div
              class="sc-bdnxRM DhlHX"
            >
              <p
                class="sc-iCoGMd gGpfF"
              >
                From Earth (C-137), identifies as Male of Human race.
              </p>
              <p
                class="sc-iCoGMd KOcH"
              >
                Current Status: Alive
              </p>
            </div>
          </article>
        </div>
      </section>
      <section
        class="sc-bdnxRM byZzmS"
        data-testid="episode-wrapper"
      >
        <h1
          class="sc-jSFjdj cYWCdr"
        >
          #90-Pilot
        </h1>
        <p
          class="sc-iCoGMd fFHScY"
        >
          This is the 01st episode in 01st session. It was aired on December 2, 2013. There are total of 1 featured characters in this episode
        </p>
        <div
          class="sc-bdnxRM dNzzIz"
        >
          <article
            class="sc-bdnxRM jDhRBX"
            scale="1.02"
            width="30%"
          >
            <span
              class=" lazy-load-image-background blur"
              style="display: inline-block; height: 50%; width: 100%;"
            >
              <img
                alt="episode character"
                height="50%"
                src="https://rickandmortyapi.com/api/character/avatar/1.jpeg"
                width="100%"
              />
            </span>
            <h2
              class="sc-eCApnc hQwxcN"
              color="#1f1f1f"
              font-style="normal"
            >
              Rick Sanchez
            </h2>
            <div
              class="sc-bdnxRM DhlHX"
            >
              <p
                class="sc-iCoGMd gGpfF"
              >
                From Earth (C-137), identifies as Male of Human race.
              </p>
              <p
                class="sc-iCoGMd KOcH"
              >
                Current Status: Alive
              </p>
            </div>
          </article>
        </div>
      </section>
    </main>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
